ibis/expr/rules.py

```diff
@@ -801,3 +807,120 @@ def _validate(self, args, i):
 
 def comparable(left, right):
     return ir.castable(left, right) or ir.castable(right, left)
+
+
+@six.add_metaclass(abc.ABCMeta)
+class TableColumnValidator():
+    @abc.abstractmethod
+    def validate(self):
+        pass
+
+
+class SubsetValidator(TableColumnValidator):
+    def __init__(self, *rules):
+        self.rules = rules
+
+    def validate(self, arg):
+        if isinstance(arg, ir.TableExpr):
+            # Check that columns match the schema first
+            for column_rule in self.rules:
+                # Members of a schema are arguments with a name
+                if not isinstance(column_rule, Argument):
+                    raise ValueError(
+                        'Arguments of subset schema must be instances of the '
+                        'Argument class (rules).')
+                if column_rule.name is None:
+                    raise ValueError(
+                        'Column rules must be named inside a table.')
+                try:
+                    # Use the argument's name to get
+                    column = arg[column_rule.name]
+                except IbisTypeError:
+                    if column_rule.optional:
+                        break
+                    else:
+                        raise IbisTypeError(
```

Comment:
This whole thing can be written much more clearly:

```python
if not column_rule.optional and column_rule.name not in arg:
    raise IbisTypeError(...)
```

Reply:
Still have to skip if optional. Will do:
'''                if column_rule.name not in arg:
                    if column_rule.optional:
                        continue
                    else:
                        raise IbisTypeError(
                            'No column with name {}.'.format(column_rule.name))
'''

Reply:
Did that ^